main.c

```diff
@@ -479,6 +550,164 @@ static void setworkdir(const char *wd)
+				fprintf(stderr, "Cannot find field with public key within generated section.\n");
+				error_number = 8;
+				break;
+			}
+			const char *const p_pub_begin = pfield_pub_begin + KEYS_FIELD_PUBLICKEY_LEN;
+			if (pendrecord - p_pub_begin < BASE64_TO_LEN(KEYS_FIELD_PUBLICKEY_LEN)) {
+				fprintf(stderr, "Generated section it too small to keep base64 encoding of public key.\n");
+				error_number = 9;
+				break;
+			}
+			char pubbuf[FORMATTED_PUBLIC_LEN];
+			if (-1 == base64_from((u8*)pubbuf, p_pub_begin, BASE64_TO_LEN(FORMATTED_PUBLIC_LEN))) {
+				fprintf(stderr, "Invalid base64 encoding of secret key.\n");
+				error_number = 10;
+				break;
+			}
+
+			char pathbuf[1024];
+			const size_t keys_directory_path_len = workdirlen + strlen(hostname);
+			if (keys_directory_path_len >= sizeof(pathbuf)) {
+				fprintf(stderr, "Keys directory path to is too long: %ld, max allowed length is %ld.\n", keys_directory_path_len, sizeof(pathbuf));
+				error_number = 11;
+				break;
+			}
+			strncpy(pathbuf, workdir, workdirlen);
+			strncpy(pathbuf + workdirlen, hostname, strlen(hostname));
+			pathbuf[keys_directory_path_len] = '\0';
+			if (-1 == createdir(pathbuf, use_secret_mode)) {
+				fprintf(stderr, "Cannot create directory \"%s\" for key files.\n", pathbuf);
+				error_number = 12;
+				break;
+			}
+
+			const size_t secretkey_filepath_len = keys_directory_path_len + PATH_SEPARATOR_LEN + strlen(secret_key_filename);
+			if (secretkey_filepath_len >= sizeof(pathbuf)) {
+				fprintf(stderr, "Path to file with secret key is too long %ld, max allowed length is %ld.\n", secretkey_filepath_len, sizeof(pathbuf));
+				error_number = 13;
+				break;
+			}
+			pathbuf[keys_directory_path_len] = '/';
+			strncpy(pathbuf + keys_directory_path_len + PATH_SEPARATOR_LEN, secret_key_filename, strlen(secret_key_filename));
+			pathbuf[secretkey_filepath_len] = '\0';
+			if (-1 == writetofile(pathbuf, (u8*)secbuf, sizeof(secbuf), use_secret_mode)) {
+				fprintf(stderr, "Can't write secret key to file \"%s\".\n", pathbuf);
+				error_number = 14;
+				break;
+			}
+
+			const size_t publickey_filepath_len = keys_directory_path_len + PATH_SEPARATOR_LEN + strlen(public_key_filename);
+			if (publickey_filepath_len >= sizeof(pathbuf)) {
+				fprintf(stderr, "Path to file with public key is too long %ld, max allowed length is %ld.\n", publickey_filepath_len, sizeof(pathbuf));
+				error_number = 15;
+				break;
+			}
+			pathbuf[keys_directory_path_len] = '/';
+			strncpy(pathbuf + keys_directory_path_len + PATH_SEPARATOR_LEN, public_key_filename, strlen(public_key_filename));
+			pathbuf[publickey_filepath_len] = '\0';
+			if (-1 == writetofile(pathbuf, (u8*)pubbuf, sizeof(pubbuf), use_public_mode)) {
+				fprintf(stderr, "Can't write public key to file \"%s\".\n", pathbuf);
+				error_number = 16;
+				break;
+			}
+
+			pathbuf[keys_directory_path_len] = '\0';
+			fprintf(stderr, "Keys successfully exported to directory \"%s\".\n", pathbuf);
```

Comment:
maybe should print to stdout too. for consistency.

Reply:
s/too/instead/
2 lines stating the same thing would be eyesore.
if printing to stdout, it should output only name without anything else just like it does in non-yaml mode.

Reply:
actually should print to fout, not stdout, to support -o

Reply:
Logging never should not go to the stream where we posting results (this breaks, for example, append mode of storing results). This is the design of most unix utils (can't find any counterexample in fact).
I suggest separate it everywhere.

Reply:
that would be not logging, but output.
same kind of output it makes in non-yaml mode.
that list of onions are not log, they intentionally go to stdout to allow utilities to parse them.
I kind of want consistency there -- if successful, print to output handle, if unsuccessful - log error.